x/inflation/types/v1beta2/inflation_calculator.go

```diff
@@ -17,74 +18,85 @@ func GetInflationCalculator(
-		powDec := sdk.NewDecFromBigInt(
-			decimal.WithPrecision(sdk.Precision).
-				Mul(pow, decimal.New(1000000, 0)).
-				Int(nil),
-		).QuoInt64(1000000)
+		return inflationCalculator(ctx.BlockTime(), genesisTime, minter, params, inflationParams, bondedRatio)
+	}
+}
 
-		idealInflation := inflationParams.InitialInflation.Mul(powDec)
+// inflationCalculator calculate current inflation value
+// - btime - block time from sdk.Context
+// - gtime - genesis time
+func inflationCalculator(btime, gtime time.Time, minter minttypes.Minter, mparams minttypes.Params, iparams Params, bondedRatio sdk.Dec) sdk.Dec {
+	// inflationDecayFactor, _ := decimal.WithPrecision(sdk.Precision).SetString(iparams.InflationDecayFactor.String())
+	inflationDecayFactor := new(decimal.Big)
+	if _, valid := inflationDecayFactor.SetString(iparams.InflationDecayFactor.String()); !valid {
+		panic(fmt.Sprintf("InflationDecayFactor contains invalid value [%s]. expected integer/float", iparams.InflationDecayFactor.String()))
+	}
 
-		// (1 - bondedRatio/GoalBonded) * InflationRateChange
-		inflationRateChangePerYear := sdk.OneDec().
-			Sub(bondedRatio.Quo(params.GoalBonded)).
-			Mul(params.InflationRateChange)
-		inflationRateChange := inflationRateChangePerYear.Quo(sdk.NewDec(int64(params.BlocksPerYear)))
+	// years passed since genesis = seconds passed since genesis / number of seconds per year
+	// can be a fraction, eg: 0.5
+	yearsPassed := decimal.WithPrecision(sdk.Precision).
+		Quo(
+			// seconds since genesis
+			decimal.WithPrecision(sdk.Precision).
+				Sub(
+					decimal.New(btime.Unix(), 0),
+					decimal.New(gtime.Unix(), 0),
+				),
+			// Number of hours in an year = 8766 (averaging the leap year hours)
+			// Number of minutes in an hour = 60
+			// Number of seconds in a minute = 60
+			// => Number of seconds per year = 60 * 60 * 8766 = 31557600
+			decimal.New(31557600, 0),
+		)
+	// 2^(-t/tHalf)
+	pow := decimal.WithPrecision(sdk.Precision)
+	pow = pow.Context.
+		Pow(
+			pow,
+			decimal.New(2, 0),
+			decimal.WithPrecision(sdk.Precision).
+				Mul(
+					decimal.New(-1, 0),
+					decimal.WithPrecision(sdk.Precision).
+						Quo(yearsPassed, inflationDecayFactor),
+				),
+		)
+	// convert pow to sdk.Dec with a 6 unit precision: sdk.Decimal(big.Int(pow * 10^6)) / 10^6
+	powDec := sdk.NewDecFromBigInt(
+		decimal.WithPrecision(sdk.Precision).
+			Mul(pow, decimal.New(1000000, 0)).
+			Int(nil),
+	).QuoInt64(1000000)
+
+	// yearsPassed :=
+	idealInflation := iparams.InitialInflation.Mul(powDec)
 
-		// note inflationRateChange may be negative
-		currentInflation := minter.Inflation.Add(inflationRateChange)
+	// (1 - bondedRatio/GoalBonded) * InflationRateChange
+	inflationRateChangePerYear := sdk.OneDec().
+		Sub(bondedRatio.Quo(mparams.GoalBonded)).
+		Mul(mparams.InflationRateChange)
+	inflationRateChange := inflationRateChangePerYear.Quo(sdk.NewDec(int64(mparams.BlocksPerYear)))
```

Comment:
Should be able to just do `sdk.NewDecFromInt(sdk.NewIntFromUint64(mparams.BlocksPerYear))` here to avoid the cast

Reply:
💯